tests/test_sig.c

```diff
@@ -118,6 +119,76 @@ static OQS_STATUS sig_test_correctness(const char *method_name) {
 		goto err;
 	}
 
+	/* testing signing with context, if supported */
+	OQS_randombytes(ctx, 257);
+	if (sig->sig_with_ctx_support) {
+		for (size_t i = 0; i < 256; ++i) {
+			rc = OQS_SIG_sign_with_ctx_str(sig, signature, &signature_len, message, message_len, ctx, i, secret_key);
+			OQS_TEST_CT_DECLASSIFY(&rc, sizeof rc);
+			if (rc != OQS_SUCCESS) {
+				fprintf(stderr, "ERROR: OQS_SIG_sign_with_ctx_str failed\n");
+				goto err;
+			}
+
+			OQS_TEST_CT_DECLASSIFY(public_key, sig->length_public_key);
+			OQS_TEST_CT_DECLASSIFY(signature, signature_len);
+			rc = OQS_SIG_verify_with_ctx_str(sig, message, message_len, signature, signature_len, ctx, i, public_key);
+			OQS_TEST_CT_DECLASSIFY(&rc, sizeof rc);
+			if (rc != OQS_SUCCESS) {
+				fprintf(stderr, "ERROR: OQS_SIG_verify_with_ctx_str failed\n");
+				goto err;
+			}
+
+			/* modify the signature to invalidate it */
+			OQS_randombytes(signature, signature_len);
```

Comment:
Pretty drastic sig change for negative test. What about only "++"ing a single random byte?

Reply:
Thank you for the suggestion! The tests were originally implemented this way. Considering the security claims of the signature schemes in liboqs, all of them assert "EUF-CMA". Would this test align with that security claim? Tagging @dstebila for input and to confirm whether the test was intentionally designed this way originally.

I’ve run the suggested test by @baentsch and noticed that the CROSS signature scheme seems to fail in some cases, passing the verification of modified signatures.

For ML-DSA (FIPS 204) at least, which claims the stronger SUF-CMA property, this test should be applicable. Section 3.1 of [FIPS 204](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.204.pdf) states:

> ML-DSA is designed to be strongly existentially unforgeable under chosen message attack (SUF-CMA).
That is, it is expected that even if an adversary can get the honest party to sign arbitrary messages, the
adversary cannot create any additional valid signatures based on the signer’s public key, including on
messages for which the signer has already provided a signature.

Side-note: I've updated the security claim of ML-DSA to SUF-CMA in the ml_dsa.yml.

Reply:
I guess technically for an EUF-CMA scheme it would not be a violation of security if you found a different signature for the same message, so for those schemes a better test would be changing bytes of the message, rather than bytes of the signature.

For SUF-CMA schemes, it would be a violation of security if a different signature works for the same message, so for those schemes changing bytes of the signature is a good test. I do think changing as few bytes as possible is better. 

Very curious that CROSS passes verification of modified signatures -- perhaps some of those bytes are padding or formatting?

Reply:
I'll defer to @dstebila's expertise, but I think both versions "test" SUF-CMA security. Given a valid message-signature pair (_m_, _σ_), it does not break EUF-CMA security to produce a second valid signature _σ'_ on the same message—whether that second signature differs in a single bit or every single bit. To break EUF-CMA security, the attacker has to produce a forgery for a message _m'_ for which it does _not_ already have a valid signature. On the other hand, SUF-CMA security is broken even if the attacker produces a second valid signature for the same message.

In order to "test" EUF-CMA security, we would have to modify the message. Modifying the signature would be optional.

I have always viewed this test as not testing "security" as much as "correctness" (in quotations because the two notions are obviously connected)—it's a sanity check that the verification algorithm doesn't trivially accept. Guessing a signature to see if it verifies is pretty much the most naïve "attack" possible, and will only catch trivially broken schemes (likely not breaking the scheme itself but rather finding an implementation error).

In principle, I do agree with @baentsch that a smaller modification is better as it's more likely to capture subtle correctness errors. However, if a scheme does not claim SUF-CMA security, we may run into false positives. I'm not sure if any of the schemes in liboqs are known to _not_ be SUF-CMA secure.

Reply:
Gah, beat me to it @dstebila!

Reply:
Thanks to both @dstebila and @SWilson4 for the reasoning, and to @baentsch for opening the related issue on CROSS: [#1995](https://github.com/open-quantum-safe/liboqs/issues/1995).

We could do the following testing strategy:

- For EUF-CMA-secure schemes: A negative test would involve modifying the message by flipping a random bit.
- For SUF-CMA-secure schemes: There would be two negative tests:
The first is the same as the EUF-CMA test (modifying the message).
The second involves modifying the signature by flipping a random bit.

To implement this, we would need an additional variable in the OQS_SIG struct, `suf_cma`, analogous to the existing `euf_cma` variable. The only scheme with documented SUF-CMA claims so far in liboqs appears to be ML-DSA.

Does this approach sound reasonable, or might it be overkill?

Reply:
Unrelated to the contents of this discussion, can I suggest moving this conversation to a new issue such as to close out this PR?

Reply:
> I'm not sure if any of the schemes in liboqs are known to not be SUF-CMA secure.

Particularly this statement by @SWilson4 below (that for some reason I cannot respond to) makes me nervous in the sense that this discussion may become "long winded" (reminds me of "unknown unknowns" :-) For now, apologies for having opened this can of worms. For now, let's keep it closed and disregard my comment and keep tests as they've "always been" but feel free to open a separate issue.

Reply:
I'm hesitant to make this test dependent on the security model because
1. it's not a particularly good test of the security of the scheme
2. even for schemes which only claim EUF-CMA, a "weak" forgery may be unexpected and worth investigation (as we just saw with #1995).

Reply:
Created issue #1997 to contain follow-up discussion.